### DISCUSSION TITLE
Type Definition of name prop in FormField Component

### DISCUSSION THREAD

**Discussion by lsbyerley** [shadcn-ui/ui#2404] [Q&A]
I am looping through an array of form fields in my form component and it is using the FormField component. For some reason, typescript compains at the name attribute in FormField. I can't seem to figure out why? Any help is appreciated

<img width="795" alt="Screenshot 2024-01-12 at 9 12 14 PM" src="https://github.com/shadcn-ui/ui/assets/3066258/2a988575-ef1b-4e76-9e8c-472e6de40c79">

File Below..

```typescript

'use client';

import { zodResolver } from '@hookform/resolvers/zod';
import { useForm } from 'react-hook-form';
import * as z from 'zod';
import { Button } from '@/components/ui/button';
import {
  Form,
  FormControl,
  FormDescription,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from '@/components/ui/form';
import { Input } from '@/components/ui/input';

interface FormFields {
  fieldName: string
  fieldLabel: string
  fieldDesc: string
}

interface props {
  createDeal: Function
}

const formFields: FormFields[] = [
  { fieldName: 'test', fieldLabel: 'Test', fieldDesc: 'Test of deal' },
  { fieldName: 'type', fieldLabel: 'Type', fieldDesc: 'Type of deal' },
] as const;

const newDealSchema = z.object({
  test: z.string(),
  type: z.string(),
})

const CreateForm = ({ createDeal }: props) => {
  const form = useForm<z.infer<typeof newDealSchema>>({
    resolver: zodResolver(newDealSchema),
    defaultValues: {
      test: '',
      type: ''
    },
  });

  const onSubmit = (values: z.infer<typeof newDealSchema>) => {
    createDeal(values);
  };

  return (
    <Form {...form}>
      <form onSubmit={form.handleSubmit(onSubmit)} className='space-y-8'>
        {formFields.map((f) => (
          <FormField
            key={f.fieldName}
            control={form.control}
            name={f.fieldName}
            render={({ field }) => (
              <FormItem>
                <FormLabel>{f.fieldLabel}</FormLabel>
                <FormControl>
                  <Input {...field} />
                </FormControl>
                <FormDescription>{f.fieldDesc}</FormDescription>
                <FormMessage />
              </FormItem>
            )}
          />
        ))}
        <Button type='submit'>Submit</Button>
      </form>
      <Button onClick={() => {
        form.reset();
        return;
      }}>Clear</Button>
    </Form>
  )
};

export default CreateForm;
```

**Comment by PieterDePauw** [comment]
In your code, the fieldName property in the FormFields interface is defined as a string. This means it can represent any string value. However, in your form schema (newDealSchema), you're using Zod to define specific field names (test and type). TypeScript infers the types from this schema as an object with keys as string literals - specifically "test" and "type".

TypeScript expects the field names to exactly match the keys in your schema. In other words, it expects these names to be the specific string literals "test" and "type", not just any string. So, in your CreateForm component, when you are mapping over formFields and passing f.fieldName as the name prop to the FormField component, TypeScript will flag an error because f.fieldName, being a general string, could potentially be _any string_ and not necessarily the specific strings "test" or "type" required by the form schema.

**Reply by lsbyerley** [reply]
@PieterDePauw Thank you for the detailed explanation. Since `fieldName` is the name attribute of a form element, it can be _any_ string. So i'm confused why it thinks it should be `test` or `type`. How can i change this to accept any string instead?

**Reply by PieterDePauw** [reply]
The error you're encountering in TypeScript is related to how types are being inferred and used in your code. Let's break down the error and your code to understand the issue:

```typescript
Type 'string' is not assignable to type '"test" | "type"'
```

This error message indicates that TypeScript expects a string literal type — either "test" or "type" — but it's receiving a more general string type instead.

You are indeed defining `fieldName` as a string in the `FormFields` interface and `formFields` array. However, in the `useForm` hook, the form is typed to expect specific keys ("test" and "type") which are explicitely inferred from the `newDealSchema` schema object. As a result, the useForm hook expect not just a string but an even more constrained type of these two string literals

When mapping over `formFields` and passing `f.fieldName` to the `name` prop of `FormField`, TypeScript cannot guarantee that `f.fieldName` is one of the specific strings "test" or "type". It only knows that it's a string, which is broader than the "test" | "type" type expected.

Basically, you have explicitly typed both the data that you will provide to `useForm`  (`FormFields`) as well as the data that `useForm` should expect (`newDealSchema`) and one is more constrained than the other.

I don't know why you have hard-coded the `formFields`, so there are a number of ways to solve this error. You could add the same constraint to the `FormFields` interface, since you actually already know all possible values for `fieldName`. You could also make use a more relaxed `newDealSchema`, such as `z.record(z.string());` (meaning any object with string values is valid). But that of course defeats the purpose of the form input validation.

Happy to help!

**Comment by lsbyerley** [comment]
Good looks, that helped a lot!
